app/src/src_nonfreenet/org/breezyweather/sources/cwa/CwaResultConverter.kt

```diff
@@ -0,0 +1,645 @@
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
+ * License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Breezy Weather. If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package org.breezyweather.sources.cwa
+
+import android.graphics.Color
+import breezyweather.domain.location.model.Location
+import breezyweather.domain.weather.model.AirQuality
+import breezyweather.domain.weather.model.Alert
+import breezyweather.domain.weather.model.AlertSeverity
+import breezyweather.domain.weather.model.Astro
+import breezyweather.domain.weather.model.Current
+import breezyweather.domain.weather.model.Daily
+import breezyweather.domain.weather.model.HalfDay
+import breezyweather.domain.weather.model.Normals
+import breezyweather.domain.weather.model.PrecipitationProbability
+import breezyweather.domain.weather.model.Temperature
+import breezyweather.domain.weather.model.UV
+import breezyweather.domain.weather.model.WeatherCode
+import breezyweather.domain.weather.model.Wind
+import breezyweather.domain.weather.wrappers.AirQualityWrapper
+import breezyweather.domain.weather.wrappers.HourlyWrapper
+import breezyweather.domain.weather.wrappers.SecondaryWeatherWrapper
+import breezyweather.domain.weather.wrappers.WeatherWrapper
+import org.breezyweather.common.exceptions.InvalidLocationException
+import org.breezyweather.common.source.SecondaryWeatherSourceFeature
+import org.breezyweather.sources.cwa.json.CwaAlertResult
+import org.breezyweather.sources.cwa.json.CwaAstroResult
+import org.breezyweather.sources.cwa.json.CwaLocationTown
+import org.breezyweather.sources.cwa.json.CwaNormalsResult
+import org.breezyweather.sources.cwa.json.CwaUvResult
+import org.breezyweather.sources.cwa.json.CwaWeatherForecast
+import org.breezyweather.sources.cwa.json.CwaWeatherResult
+import java.text.SimpleDateFormat
+import java.util.Calendar
+import java.util.Locale
+import java.util.TimeZone
+
+fun convert(
+    location: Location,
+    town: CwaLocationTown
+): Location {
+    return location.copy(
+        timeZone = "Asia/Taipei",
+        country = "臺灣",
+        province = town.ctyName,
+        city = town.townName!!,
+        district = town.villageName
+    )
+}
+
+fun convert(
+    weatherResult: CwaWeatherResult,
+    uvResult: CwaUvResult,
+    normalsResult: CwaNormalsResult,
+    alertResult: CwaAlertResult,
+    sunResult: CwaAstroResult,
+    moonResult: CwaAstroResult,
+    location: Location,
+    id: String,
+    ignoreFeatures: List<SecondaryWeatherSourceFeature>
+): WeatherWrapper {
+    return WeatherWrapper(
+        current = getCurrent(weatherResult, uvResult, ignoreFeatures),
+        normals = if (!ignoreFeatures.contains(SecondaryWeatherSourceFeature.FEATURE_NORMALS)) getNormals(normalsResult) else null,
+        dailyForecast = getDailyForecast(weatherResult.data!!.aqi[0].town!!.daily, sunResult, moonResult),
```

Comment:
Please safely throw an exception instead forcing Null Pointer Exceptions. You can see an example here: https://github.com/breezy-weather/breezy-weather/blob/v5.2.4/app/src/main/java/org/breezyweather/sources/openmeteo/OpenMeteoResultConverter.kt#L94-L97

Reply:
Will do! Thanks for the pointer.